EIPS/eip-3540.md

```diff
@@ -93,88 +89,94 @@ The list of section headers is terminated with the *section headers terminator b
-   - Exactly one data section header MUST be present immediately following the code section.
-2. Any version other than `0x01` is invalid.
+The following validity constraints are placed on the container format:
 
-(*Remark:* Contract creation code SHOULD set the section size of the data section so that the constructor arguments fit it.)
+- minimum valid header size is `15` bytes
+- `version` must be `0x01`
+- `types_size` is divisible by `4`
+- the number of code sections must be equal to `types_size / 4`
+- the number of code sections must not exceed 1024
+- `code_size` may not be 0
+- the number of container sections must not exceed 256
+- `container_size` may not be 0, but container sections are optional
+- the total size of a deployed container without container sections must be `13 + 2*num_code_sections + types_size + code_size[0] + ... + code_size[num_code_sections-1] + data_size`
+- the total size of a deployed container with at least one container section must be `16 + 2*num_code_sections + types_size + code_size[0] + ... + code_size[num_code_sections-1] + data_size + 2*num_container_sections + container_size[0] + ... + container_size[num_container_sections-1]`
+- the total size of not yet deployed container might be up to `data_size` lower than the above values due to how the data section is rewritten and resized during creation
 
 ### Changes to execution semantics
 
-For clarity, the *container* refers to the complete account code, while *code* refers to the contents of the code section only.
-
-1. Execution starts at the first byte of the first code section, and PC is set to 0.
-2. Execution stops if `PC` goes outside the code section bounds.
-3. `PC` returns the current position within the *code*.
-4. `CODECOPY`/`CODESIZE`/`EXTCODECOPY`/`EXTCODESIZE`/`EXTCODEHASH` keeps operating on the entire *container*.
-5. The input to `CREATE`/`CREATE2` is still the entire *container*.
-6. The size limit for deployed code as specified in [EIP-170](./eip-170.md) and for initcode as specified in [EIP-3860](./eip-3860.md) is applied to the entire *container* size, not to the *code* size. This also means if initcode validation fails, it is still charged the EIP-3860 `initcode_cost`.
-7. When an EOF1 contract performs a `DELEGATECALL` the target must be EOF1. If it is not EOF1, the `DELEGATECALL` execution finishes as a failed call by pushing `0` to the stack. Only initial gas cost of `DELEGATECALL` is consumed (similarly to the call depth check) and the target address still becomes warm.
-
-(*Remark:* Due to [EIP-4750](./eip-4750.md), `JUMP` and `JUMPI` are disabled and therefore are not discussed in relation to EOF.)
+For an EOF contract:
+- Execution starts at the first byte of code section 0, and `pc` is set to 0.
+- `pc` is scoped to the executing code section
+- `CODESIZE`, `CODECOPY`, `EXTCODESIZE`, `EXTCODECOPY`, `EXTCODEHASH`, `GAS` are rejected by validation in EOF contracts, with no replacements
```

Comment:
I think we should either remove this, or have a full list of deprecated instructions. But it should be in EIP-3670, so maybe remove.

Reply:
After some deliberation, I decided to have the deprecations accompany the replacement definitions wherever possible, and here list all the deprecations without replacements. The EIPs also kinda followed this model, so retaining this should give a smaller diff. But mostly, I think it reads easier, if the legacy and EOF logical counterparts are nearby.

But that's not a strong opinion, I spent quite a lot of time juggling these alternatives, all have pros and cons